### PR TITLE
fix: add tls-ca-cert-file flag to allow tls certificate validation

### PR DESCRIPTION
**Pull request by kostasrim**
1. add `tls-ca-cert-file` flag to allow loading of the CA cert file
2. add `tls-ca-cert-dir` flag to allow loading of the CA files from directories
3. enables redis-cli to connect over tls without --insecure flag #1418 (addresses item 3 in the list)

### CODE REVIEW

**Comment by royjacobson** [comment]
Ahhhhhh Ok now I understand what this is supposed to do and I don't like it :laughing: 

Usually the trusted CA certificates are just supplied by the operating system. [This](https://www.happyassassin.net/posts/2015/01/12/a-note-about-ssltls-trusted-certificate-stores-and-platforms/) blog suggests doing that by `SSL_CTX_set_default_verify_paths(ctx)`.

**Comment by kostasrim** [comment]
@royjacobson 
>So really, what you should do - like I said - is first of all, try letting OpenSSL handle it. If you're using OpenSSL directly, the function you want is SSL_CTX_set_default_verify_paths(). If a default trust store was specified at the time the OpenSSL build your app winds up using was done, it'll get used

That really doesn't sound good and it can lead to surprising behavior. When I start my system, I want to be in control of which CA I trust and some random folder which can provide an extra gateway that is beyond by control sound clumsy and error prone at best. Besides, I can pass the default folder when I start my system, and that way I am `explicit` to what CA I trust. 

Moreover, redis does the same: https://github.com/redis/redis/blob/unstable/src/tls.c#L257 and from a UX experience I would like to stick as close to that as possible.

File: tests/dragonfly/connection_test.py [hunk 1]
```diff
@@ -419,18 +420,68 @@ async def test_large_cmd(async_client: aioredis.Redis):
...
 
     # Try to connect on master on admin port
     client = aioredis.Redis(port=master.admin_port)
     assert await client.ping()
+    await client.close()
+
+
+@pytest.mark.asyncio
+async def test_tls_insecure(with_ca_tls_server_args, with_tls_client_args, df_local_factory):
+    master = df_local_factory.create(port=BASE_PORT, **with_ca_tls_server_args)
```
Comment (romange) [inline tests/dragonfly/connection_test.py:442]: why do you need to specify `BASE_PORT` ?

Comment (kostasrim) [reply tests/dragonfly/connection_test.py:442]: not a particular reason, it's a convention I seen throughout the tests and I thought I should follow it

File: tests/dragonfly/connection_test.py [hunk 2]
```diff
@@ -419,18 +420,68 @@ async def test_large_cmd(async_client: aioredis.Redis):
...
+
+
+@pytest.mark.asyncio
+async def test_tls_full_auth(with_ca_tls_server_args, with_ca_tls_client_args, df_local_factory):
+    master = df_local_factory.create(port=BASE_PORT, **with_ca_tls_server_args)
+    master.start()
+
+    # Try to connect on master on admin port
+    client = aioredis.Redis(port=master.port, **with_ca_tls_client_args)
+    assert await client.ping()
```
Comment (kostasrim) [inline tests/dragonfly/connection_test.py:458]: `ping` is a BUG. Always succeeds. Generally speaking, avoid it.

File: tests/dragonfly/conftest.py [hunk 3]
```diff
@@ -232,7 +232,7 @@ def memcached_connection(df_server: DflyInstance):
 def gen_tls_cert(df_factory: DflyInstanceFactory):
     tls_server_key_file_name = "df-key.pem"
     tls_server_cert_file_name = "df-cert.pem"
-    dfly_path = df_factory.dfly_path
+    dfly_path = df_factory.dfly_path + "/"
```
Comment (adiholden) [inline tests/dragonfly/conftest.py:235]: use os.path.join when you want to create the full path from file name and directory

File: src/facade/dragonfly_listener.cc [hunk 4]
```diff
@@ -77,19 +80,24 @@ static SSL_CTX* CreateSslCntx() {
     if (!tls_cert_file.empty()) {
       // TO connect with redis-cli you need both tls-key-file and tls-cert-file
       // loaded. Use `redis-cli --tls -p 6380 --insecure  PING` to test
-
       CHECK_EQ(1, SSL_CTX_use_certificate_chain_file(ctx, tls_cert_file.c_str()));
     }
+
+    const auto tls_ca_cert_file = GetFlag(FLAGS_tls_ca_cert_file);
```
Comment (adiholden) [inline src/facade/dragonfly_listener.cc:86]: Why did you add this?
Did we have a task on this or you just saw this was missing and wanted to add it ?
I see redis has tls-auth-clients config, if this config is set then validation is done.
I suggest adding tls_auth_client flag and if it is true you should make sure that one of the flags dir / file is set

File: tests/dragonfly/connection_test.py [hunk 5]
```diff
@@ -419,18 +420,61 @@ async def test_large_cmd(async_client: aioredis.Redis):
 
 
 @pytest.mark.asyncio
-async def test_reject_non_tls_connections_on_tls_master(with_tls_args, df_local_factory):
-    master = df_local_factory.create(admin_port=1111, port=1211, **with_tls_args)
+async def test_reject_non_tls_connections_on_tls_master(with_tls_server_args, df_local_factory):
+    master = df_local_factory.create(admin_port=1111, port=1211, **with_tls_server_args)
```
Comment (adiholden) [inline tests/dragonfly/connection_test.py:424]: rename master to server

File: tests/dragonfly/connection_test.py [hunk 6]
```diff
@@ -419,18 +420,61 @@ async def test_large_cmd(async_client: aioredis.Redis):
 
 
 @pytest.mark.asyncio
-async def test_reject_non_tls_connections_on_tls_master(with_tls_args, df_local_factory):
-    master = df_local_factory.create(admin_port=1111, port=1211, **with_tls_args)
+async def test_reject_non_tls_connections_on_tls_master(with_tls_server_args, df_local_factory):
```
Comment (adiholden) [inline tests/dragonfly/connection_test.py:423]: remove master from the test name

File: tests/dragonfly/conftest.py [hunk 7]
```diff
@@ -257,14 +257,54 @@ def gen_tls_cert(df_factory: DflyInstanceFactory):
...
+    step4 = rf'openssl req -newkey rsa:4096 -nodes -keyout {tls_client_key} -out {tls_client_req} -subj "/C=It/ST=RM/L=Rome/O=KK/OU=Comp/CN=Gr/emailAddress=random@gmail.com"'
+    subprocess.run(step4, shell=True)
+
+    tls_client_cert_file_name = "client-cert.pem"
+    tls_client_cert = dfly_path + tls_client_cert_file_name
+    step5 = fr'openssl x509 -req -in {tls_client_req} -days 1 -CA {ca_cert} -CAkey {ca_key} -CAcreateserial -out {tls_client_cert}'
+
+    subprocess.run(step5, shell=True)
+
+    return [tls_server_key, tls_server_cert, ca_cert, tls_client_key, tls_client_cert]
```
Comment (adiholden) [inline tests/dragonfly/conftest.py:276]: I find the current design a bit confusing
I suggest creating a fixture called gen_ca_cert which will return {"ssl_ca_certs", <path>}
create a function gen_tls_cert (not a fixture)
create a fixture called gen_server_cert which will return {"tls_key_file", <path> , "tls_cert_file", <path>}
and a fixture called gen_client_cert which will return {"ssl_keyfile", <path>, "ssl_certfile" , path}

In the test where you need to pass no_tls_on_admin_port , this will be done inside the test and not from the fixture
I think with the suggested design it will be more clear which args are passed to each test case

File: src/facade/dragonfly_listener.cc (hunk 4 above)
Comment (romange) [reply src/facade/dragonfly_listener.cc:86]: if both flags are empty, it's equivalent to `tls_auth_client=false` 
if some of the flags are not empty, it's equivalent to `tls_auth_client=true` 

Why do you need the state where `tls_auth_client=false` and some of the flags are not empty?

Comment (adiholden) [reply src/facade/dragonfly_listener.cc:86]: just have similar config as redis

File: tests/dragonfly/conftest.py [hunk 8]
```diff
@@ -229,42 +230,89 @@ def memcached_connection(df_server: DflyInstance):
...
-    subprocess.run(step3, shell=True)
-    return tls_server_key_file_name, tls_server_cert_file_name
+    step2 = fr'openssl x509 -req -in {tls_server_req} -days 1 -CA {ca_cert} -CAkey {ca_key} -CAcreateserial -out {tls_server_cert}'
+    subprocess.run(step2, shell=True)
+
+    return {"server_key": tls_server_key, "server_cert": tls_server_cert}
+
+
+@pytest.fixture(scope="session")
+def gen_client_cert(df_factory: DflyInstanceFactory, gen_ca_cert):
```
Comment (adiholden) [inline tests/dragonfly/conftest.py:271]: Is there a real difference between gen_client_cert and gen_server_cert functions?or is it just the file names ? 
If they are the same I think you should create a function gen_cert which you will pass the filenames

File: tests/dragonfly/conftest.py [hunk 9]
```diff
@@ -229,42 +230,89 @@ def memcached_connection(df_server: DflyInstance):
...
 
 @pytest.fixture(scope="session")
-def with_tls_args(df_factory: DflyInstanceFactory, gen_tls_cert):
-    tls_server_key_file_name, tls_server_cert_file_name = gen_tls_cert
+def with_tls_server_args(gen_server_cert):
     args = {"tls": "",
-            "tls_key_file": df_factory.dfly_path + tls_server_key_file_name,
-            "tls_cert_file": df_factory.dfly_path + tls_server_cert_file_name,
-            "no_tls_on_admin_port": "true"}
+            "tls_key_file": gen_server_cert["server_key"],
```
Comment (adiholden) [inline tests/dragonfly/conftest.py:294]: as discussed, this fixture is redundant as this args can be returned from gen_client_cert fixture

File: src/facade/dragonfly_listener.cc (hunk 4 above)
Comment (kostasrim) [reply src/facade/dragonfly_listener.cc:86]: yes but in this particular case it's a redundant flag. I am all in when it comes to be as close to redis as possible but when it's tangible to improve I opt in for that.

File: tests/dragonfly/__init__.py [hunk 10]
```diff
@@ -162,7 +162,8 @@ def __str__(self):
 
     @property
     def dfly_path(self):
-        return str(os.path.dirname(self.params.path))
+         path = os.path.dirname(self.params.path)
+         return str(os.path.join(path, ''))
```
Comment (adiholden) [inline tests/dragonfly/__init__.py:166]: On the previous comment I meant that you should use the os.path.join when you already have the filename

File: tests/dragonfly/conftest.py [hunk 11]
```diff
@@ -229,42 +230,65 @@ def memcached_connection(df_server: DflyInstance):
...
+    subprocess.run(step, shell=True)
+
+    return {"ca_key": ca_key, "ca_cert": ca_cert}
+
 
-    # Step 2
-    # Generate Dragonfly's private key and certificate signing request (CSR)
+@pytest.fixture(scope="session")
+def with_tls_server_args(df_factory: DflyInstanceFactory, gen_ca_cert):
+    dfly_path = df_factory.dfly_path
```
Comment (adiholden) [inline tests/dragonfly/conftest.py:251]: actually I think it makes more sense here to use the tmp dir.
df_factory.tmp_dir

File: tests/dragonfly/conftest.py [hunk 12]
```diff
@@ -229,42 +230,65 @@ def memcached_connection(df_server: DflyInstance):
...
+    tls_server_key_file_name = "df-key.pem"
     tls_server_key = dfly_path + tls_server_key_file_name
     tls_server_req = dfly_path + "df-req.pem"
-    step2 = rf'openssl req -newkey rsa:4096 -nodes -keyout {tls_server_key} -out {tls_server_req} -subj "/C=GR/ST=SKG/L=Thessaloniki/O=KK/OU=Comp/CN=Gr/emailAddress=does_not_exist@gmail.com"'
-    subprocess.run(step2, shell=True)
-
-    # Step 3
-    # Use CA's private key to sign dragonfly's CSR and get back the signed certificate
+    tls_server_cert_file_name = "df-cert.pem"
```
Comment (adiholden) [inline tests/dragonfly/conftest.py:255]: tls_server_req = os.path.join(df_factory.tmp_dir, "df-req.pem")
tls_server_key =  os.path.join(df_factory.tmp_dir, "df-key.pem")
tls_server_cert = os.path.join(df_factory.tmp_dir, "df-cert.pem")

File: tests/dragonfly/conftest.py [hunk 13]
```diff
@@ -229,42 +230,65 @@ def memcached_connection(df_server: DflyInstance):
...
+def gen_ca_cert(df_factory: DflyInstanceFactory):
     dfly_path = df_factory.dfly_path
     # We first need to generate the tls certificates to be used by the server
 
-    # Step 1
     # Generate CA (certificate authority) key and self-signed certificate
     # In production, CA should be generated by a third party authority
     # Expires in one day and is not encrtypted (-nodes)
     # X.509 format for the key
     ca_key = dfly_path + "ca-key.pem"
```
Comment (adiholden) [inline tests/dragonfly/conftest.py:241]: ca_key = os.path.join(df_factory.tmp_dir, "ca-key.pem")
ca_cert = os.path.join(df_factory.tmp_dir, "ca-cert.pem")

File: tests/dragonfly/utility.py [hunk 14]
```diff
@@ -560,3 +561,15 @@ def create(self, **kwargs):
 
 async def disconnect_clients(*clients):
     await asyncio.gather(*(c.connection_pool.disconnect() for c in clients))
+
+
+def gen_cert(ca, req, key, cert):
```
Comment (adiholden) [inline tests/dragonfly/utility.py:566]: use more descriptive names
gen_cert rename to gen_certificate
key rename to private_key_path
cert rename to certificate_path
req rename to certificate_request_path
also in this function I think it will be more readable if you path cs_key_path and cs_certificate_path and not ca dictionary

**Review by adiholden — APPROVED** [review]
